src/api/objects/InMemorySearchProvider.js

```diff
@@ -0,0 +1,290 @@
+        const rootObject = this.openmct.objects.rootProvider.rootObject;
+        this.scheduleForIndexing(rootObject.identifier);
+        this.worker = this.startSharedWorker();
+    }
+
+    /**
+     * @private
+     */
+    getIntermediateResponse() {
+        let intermediateResponse = {};
+        intermediateResponse.promise = new Promise(function (resolve, reject) {
+            intermediateResponse.resolve = resolve;
+            intermediateResponse.reject = reject;
+        });
+
+        return intermediateResponse;
+    }
+
+    /**
+     * Query the search provider for results.
+     *
+     * @param {String} input the string to search by.
+     * @param {Number} maxResults max number of results to return.
+     * @returns {Promise} a promise for a modelResults object.
+     */
+    query(input, maxResults) {
+        if (!maxResults) {
+            maxResults = this.DEFAULT_MAX_RESULTS;
+        }
+
+        const queryId = this.dispatchSearch(input, maxResults);
+        const pendingQuery = this.getIntermediateResponse();
+        this.pendingQueries[queryId] = pendingQuery;
+
+        return pendingQuery.promise;
+    }
+
+    /**
+     * Handle messages from the worker.  Only really knows how to handle search
+     * results, which are parsed, transformed into a modelResult object, which
+     * is used to resolve the corresponding promise.
+     * @private
+     */
+    async onWorkerMessage(event) {
+        // (`🖲 Received event from search worker 🖲`, event);
+        if (event.data.request !== 'search') {
+            return;
+        }
+
+        const pendingQuery = this.pendingQueries[event.data.queryId];
+        const modelResults = {
+            total: event.data.total
+        };
+        modelResults.hits = await Promise.all(event.data.results.map(async (hit) => {
+            const identifier = this.openmct.objects.parseKeyString(hit.keyString);
+            const domainObject = await this.openmct.objects.get(identifier.key);
+
+            return domainObject;
+        }));
+
+        pendingQuery.resolve(modelResults);
+        // console.debug(`🖲 Returning model results 🖲`, modelResults);
+        delete this.pendingQueries[event.data.queryId];
+    }
+
+    /**
+     * Handle error messages from the worker.
+     * @private
+     */
+    onWorkerMessageError(event) {
+        console.error('⚙️ Error message from InMemorySearch worker ⚙️', event);
+    }
+
+    /**
+     * Handle errors from the worker.
+     * @private
+     */
+    onWorkerError(event) {
+        console.error('⚙️ Error with InMemorySearch worker ⚙️', event);
+    }
+
+    /**
+     * @private
+     */
+    startSharedWorker() {
+        // eslint-disable-next-line no-undef
+        const sharedWorkerURL = `${this.openmct.getAssetPath()}${__OPENMCT_ROOT_RELATIVE__}inMemorySearchWorker.js`;
+
+        const sharedWorker = new SharedWorker(sharedWorkerURL, 'InMemorySearch Shared Worker');
```

Comment:
I don't know how shared workers...err...work. Is this idempotent? Will it only create a new Shared Worker if one has not already been created by another tab/window?

Reply:
Per:
https://www.w3.org/TR/2021/NOTE-workers-20210128/#sharedworker
(and if I'm reading this correctly)
> If name is not the empty string and there exists a SharedWorkerGlobalScope object whose closing flag is false, whose name attribute is exactly equal to name, and that is the global object specified by an environment settings object that specifies as its origin the same origin as the origin of scriptURL, then let worker global scope be that SharedWorkerGlobalScope object.

I think yes, it's idempotent. This is also my experience at runtime in that only one worker is being created no matter how many tabs/windows I open:
<img width="1519" alt="Screen Shot 2021-12-13 at 3 48 55 PM" src="https://user-images.githubusercontent.com/9853862/145835157-7c6de426-abb4-4b3f-acc1-d1210378e044.png">
Also, if I search for example "tep" in both tabs, both interact with the same shared worker:
<img width="1403" alt="Screen Shot 2021-12-13 at 3 57 57 PM" src="https://user-images.githubusercontent.com/9853862/145835406-779b12ca-3f01-4c40-be14-67c7c0066106.png">
I think this is what we want!